익명 사용자 정보 조회 API 오류 해결

## 작업 요약
익명 사용자 정보 조회 API 오류 해결

### 문제
1.2a 버전 배포 후, access token 만료되었을 때 403 페이지 뜨는 문제 발생.

```
2025-03-09 02:50:27.233 [http-nio-8080-exec-4] [ERROR] o.a.c.c.C.[.[.[.[dispatcherServlet]: Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed: io.jsonwebtoken.ExpiredJwtException: JWT expired at 2025-03-08T11:22:06Z. Current time: 2025-03-08T17:50:27Z, a difference of 23301229 milliseconds.  Allowed clock skew: 0 milliseconds.] with root cause
io.jsonwebtoken.ExpiredJwtException: JWT expired at 2025-03-08T11:22:06Z. Current time: 2025-03-08T17:50:27Z, a difference of 23301229 milliseconds.  Allowed clock skew: 0 milliseconds.
	at io.jsonwebtoken.impl.DefaultJwtParser.parse(DefaultJwtParser.java:427)
	at io.jsonwebtoken.impl.DefaultJwtParser.parse(DefaultJwtParser.java:529)
	at io.jsonwebtoken.impl.DefaultJwtParser.parseClaimsJws(DefaultJwtParser.java:589)
	at io.jsonwebtoken.impl.ImmutableJwtParser.parseClaimsJws(ImmutableJwtParser.java:173)
	at com.exchangediary.member.service.JwtService.extractAllClaims(JwtService.java:105)
	at com.exchangediary.member.service.JwtService.extractMemberId(JwtService.java:67)
	at com.exchangediary.member.service.AnonymousService.viewAnonymousInfo(AnonymousService.java:21)
	at com.exchangediary.member.ui.ApiAnonymousController.viewAnonymousInfo(ApiAnonymousController.java:25)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:255)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:188)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:118)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:926)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:831)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1089)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:979)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:903)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:564)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:195)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at com.exchangediary.global.config.log.filter.ApiLoggingFilter.doFilter(ApiLoggingFilter.java:44)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:483)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:115)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:344)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:384)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:904)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1741)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1190)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

### 원인 파악
access token 만료 시 발생하는 ExpiredJwtException이 내가 원하는대로 처리가 안 되고 있는 것 같음


### 문제 해결

--- a/src/main/java/com/exchangediary/global/config/web/WebConfig.java
+++ b/src/main/java/com/exchangediary/global/config/web/WebConfig.java
@@ -34,8 +34,8 @@ public class WebConfig implements WebMvcConfigurer {
     @Override
     public void addInterceptors(InterceptorRegistry registry) {
         registry.addInterceptor(new JwtAuthenticationInterceptor(jwtService, cookieService, memberQueryService))
-                .addPathPatterns("/groups", "/diaries/**", "/groups/**", "/api/**")
-                .excludePathPatterns("/", "/login", "/api/kakao/callback", "/api/anonymous/info");
+                .addPathPatterns("/login", "/groups/**", "/diaries/**", "/api/**")
+                .excludePathPatterns("/", "/api/kakao/callback", "/api/anonymous/info");
         registry.addInterceptor(new GroupAuthorizationInterceptor(memberQueryService))
                 .addPathPatterns("/groups/**", "/api/groups/*/**")
                 .excludePathPatterns(

--- a/src/main/java/com/exchangediary/global/config/web/interceptor/JwtAuthenticationInterceptor.java
+++ b/src/main/java/com/exchangediary/global/config/web/interceptor/JwtAuthenticationInterceptor.java
@@ -41,6 +41,11 @@ public class JwtAuthenticationInterceptor implements HandlerInterceptor {
             response.sendRedirect("/login");
             return false;
         }
+
+        if (request.getRequestURI().equals("/login")) {
+            response.sendRedirect("/");
+            return false;
+        }
         return true;
     }
 

--- a/src/main/java/com/exchangediary/global/config/web/interceptor/JwtAuthenticationInterceptor.java
+++ b/src/main/java/com/exchangediary/global/config/web/interceptor/JwtAuthenticationInterceptor.java
@@ -25,28 +25,42 @@ public class JwtAuthenticationInterceptor implements HandlerInterceptor {
             HttpServletResponse response,
             Object handler
     ) throws IOException {
-        try {
-            String token = getJwtTokenFromCookies(request);
-            String newToken = jwtService.verifyAccessToken(token);
-
-            if (newToken != null) {
-                cookieService.addCookie(jwtService.COOKIE_NAME, newToken, response);
-                token = newToken;
-            }
-
-            Long memberId = jwtService.extractMemberId(token);
-            checkMemberExists(memberId);
-            request.setAttribute("memberId", memberId);
-        } catch (UnauthorizedException exception) {
-            response.sendRedirect("/login");
-            return false;
-        }
-
         if (request.getRequestURI().equals("/login")) {
+            try {
+                verifyJwtToken(request, response);
+            } catch (UnauthorizedException exception) {
+                return true;
+            }
             response.sendRedirect("/");
             return false;
         }
+
+        try {
+            verifyJwtToken(request, response);
+        } catch (UnauthorizedException exception) {
+            if (request.getRequestURI().startsWith("/api")) {
+                throw exception;
+            }
+            response.sendRedirect("/login");
+            return false;
+        }
         return true;
+    }
+
+    private void verifyJwtToken(HttpServletRequest request,
+                                HttpServletResponse response
+    ) throws UnauthorizedException {
+        String token = getJwtTokenFromCookies(request);
+        String newToken = jwtService.verifyAccessToken(token);
+
+        if (newToken != null) {
+            cookieService.addCookie(jwtService.COOKIE_NAME, newToken, response);
+            token = newToken;
+        }
+
+        Long memberId = jwtService.extractMemberId(token);
+        checkMemberExists(memberId);
+        request.setAttribute("memberId", memberId);
     }
 
     private String getJwtTokenFromCookies(HttpServletRequest request) {

--- a/src/main/java/com/exchangediary/member/service/JwtService.java
+++ b/src/main/java/com/exchangediary/member/service/JwtService.java
@@ -5,6 +5,7 @@ import com.exchangediary.global.exception.serviceexception.UnauthorizedException
 import com.exchangediary.member.domain.entity.RefreshToken;
 import io.jsonwebtoken.Claims;
 import io.jsonwebtoken.ExpiredJwtException;
+import io.jsonwebtoken.Jwt;
 import io.jsonwebtoken.JwtException;
 import io.jsonwebtoken.Jwts;
 import io.jsonwebtoken.SignatureAlgorithm;
@@ -52,7 +53,7 @@ public class JwtService {
                 .compact();
     }
 
-    public String verifyAccessToken(String token) {
+    public String verifyAccessToken(String token) throws UnauthorizedException {
         try {
             verifyToken(token);
         } catch (ExpiredJwtException exception) {
@@ -64,16 +65,19 @@ public class JwtService {
     }
 
     public Long extractMemberId(String token) {
-        String sub = extractAllClaims(token).getSubject();
-        return Long.valueOf(sub);
+        try {
+            String sub = extractAllClaims(token).getSubject();
+            return Long.valueOf(sub);
+        } catch (JwtException | IllegalArgumentException ignored) {}
+        return null;
     }
 
-    private void verifyRefreshToken(Long memberId) {
+    private void verifyRefreshToken(Long memberId) throws UnauthorizedException{
         RefreshToken refreshToken = refreshTokenService.findRefreshTokenByMemberId(memberId);
 
         try {
             verifyToken(refreshToken.getToken());
-        } catch (ExpiredJwtException exception) {
+        } catch (JwtException exception) {
             refreshTokenService.expireRefreshToken(refreshToken);
             throw new UnauthorizedException(
                     ErrorCode.EXPIRED_TOKEN,
@@ -97,7 +101,7 @@ public class JwtService {
         }
     }
 
-    private Claims extractAllClaims(String token) {
+    private Claims extractAllClaims(String token) throws JwtException {
         return Jwts
                 .parserBuilder()
                 .setSigningKey(getSigningKey())

--- a/src/main/java/com/exchangediary/member/ui/ApiKakaoLoginController.java
+++ b/src/main/java/com/exchangediary/member/ui/ApiKakaoLoginController.java
@@ -4,17 +4,16 @@ import com.exchangediary.member.service.CookieService;
 import com.exchangediary.member.service.JwtService;
 import com.exchangediary.member.service.KakaoService;
 import com.exchangediary.member.service.MemberRegistrationService;
-import jakarta.servlet.http.Cookie;
 import jakarta.servlet.http.HttpServletResponse;
 import lombok.RequiredArgsConstructor;
 import org.springframework.http.HttpStatus;
 import org.springframework.http.ResponseEntity;
-import org.springframework.stereotype.Controller;
 import org.springframework.web.bind.annotation.GetMapping;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
 
-@Controller
+@RestController
 @RequiredArgsConstructor
 @RequestMapping("/api/kakao")
 public class ApiKakaoLoginController {

--- a/src/main/resources/application.properties
+++ b/src/main/resources/application.properties
@@ -27,10 +27,10 @@ spring.jpa.properties.hibernate.globally_quoted_identifiers=true
 kakao.client_id=${KAKAO_CLIENT_ID}
 kakao.redirect_uri=http://localhost:8080${KAKAO_REDIRECT_URI}
 
-# Jwt
+# Jwt (access - 1hour, refresh - 30days)
 security.jwt.secret-key=${JWT_SECRET_KEY}
 security.jwt.access-token.expiration-time=3600000
-security.jwt.refresh-token.expiration-time=2149200000
+security.jwt.refresh-token.expiration-time=259200000
 
 cookie.max-age.millisecond=604800000
 

